src/core/ext/filters/deadline/deadline_filter.cc

```diff
@@ -360,7 +361,16 @@ const grpc_channel_filter grpc_client_deadline_filter = {
 
 const grpc_channel_filter grpc_server_deadline_filter = {
     deadline_server_start_transport_stream_op_batch,
-    nullptr,
+    [](grpc_channel_element*, grpc_core::CallArgs call_args,
```

Comment:
If we set the make-promise function here, doesn't it mean that the promise-based filter code will try to use this filter in promise mode even if not all of the filters in the stack are promise-based, in which case we won't be using `PromiseBasedCall`?  In that case, won't `CallContext` not actually have a pointer to `PromiseBasedCall`, in which case the call to `UpdateDeadline()` will crash?

I'm guessing that there's something I'm missing here.

Reply:
1. The promise based filter code builds a complete filter vtable - it takes the `FilterType::Create` function and uses it to supply both the make_call_promise member, but also polyfills for the traditional call_element pieces of the filter vtable (and for the channel_element pieces too) - the return value of MakePromiseBasedFilter is a grpc_channel_filter vtable with most of the functions auto generated.

2. The promise based call code looks at the vtables of all selected filters, and if they all supply make_call_promise (and the correct experiment is enabled) it chooses to build a promise based call.

So one way of getting a filter vtable with a make_call_promise is by using promise_based_filter, and often we've used that.

Another way is to hand code the filter vtable - and for a few select filters that's what we're doing - mostly those where the fact that we've coded the originals as filters is of questionable value (since they're more part of the infrastructure - say connected_channel), or where we want dramatically different behavior between the batch based stack and the promise based stack (as here, where a different context is in play).

Reply:
Ah, right -- I had forgotten that if we're manually crafting make_call_promise in this case, we're not actually using the promise-based filter code, so it won't be auto-selected, which means it won't be used unless all filters in the stack are promise-based.

Sorry for the confusion.